File: engine/strategies.py
"""Métricas por combinación y puntuación por estrategia.

Generalizado: ya no existe una "ventana laboral" fija. El estudiante define
bloqueos con dos niveles de prioridad:

  - nivel "imposible": tiempo intocable. Se filtra en el solver (una
    combinación que lo pise ni siquiera existe), no aquí.
  - nivel "evitar": "mejor si no se usa". Se mide en `minutos_en_evitar` y
    TODAS las estrategias lo minimizan como primer criterio; el sabor de cada
    estrategia decide entre las combinaciones que empatan.

Todas las llaves de puntuación son "mayor es mejor" (se ordena reverse=True).
"""
from __future__ import annotations

from .models import DIAS_LABORALES

# Franja del día usada para calcular huecos/bloques libres (no es una
# restricción, solo el lienzo de las métricas).
DIA_INICIO_MIN = 7 * 60
DIA_FIN_MIN = 19 * 60


def _minutos_traslape(sesion, bloqueos) -> int:
    """Minutos-semana en que `sesion` pisa la lista de bloqueos."""
    total = 0
    for b in bloqueos:
        dias_comunes = len(sesion.dias & b.dias)
        if dias_comunes:
            solapa = max(0, min(sesion.fin_min, b.fin_min) - max(sesion.inicio_min, b.inicio_min))
            total += solapa * dias_comunes
    return total


def compute_metrics(combinacion: list, bloqueos_evitar: list = ()) -> dict:
    sesiones_por_dia = {d: [] for d in DIAS_LABORALES}
    usa_sabado = False
    usa_domingo = False
    minutos_en_evitar = 0

    for _, opcion in combinacion:
        for s in opcion.sesiones:
            minutos_en_evitar += _minutos_traslape(s, bloqueos_evitar)
            for d in s.dias:
                if d in sesiones_por_dia:
                    sesiones_por_dia[d].append((s.inicio_min, s.fin_min))
                elif d == "SA":
                    usa_sabado = True
                elif d == "DO":
                    usa_domingo = True

    dias_con_clase = {d for d in DIAS_LABORALES if sesiones_por_dia[d]}
    dias_libres = set(DIAS_LABORALES) - dias_con_clase

    def bloque_libre_mas_largo(dia):
        ocupado = sorted(sesiones_por_dia[dia])
        libres = []
        cursor = DIA_INICIO_MIN
        for (ini, fin) in ocupado:
            ini_c, fin_c = max(ini, DIA_INICIO_MIN), min(fin, DIA_FIN_MIN)
            if ini_c > cursor:
                libres.append(ini_c - cursor)
            cursor = max(cursor, fin_c)
        if cursor < DIA_FIN_MIN:
            libres.append(DIA_FIN_MIN - cursor)
        return max(libres) if libres else (DIA_FIN_MIN - DIA_INICIO_MIN)

    bloques = {d: bloque_libre_mas_largo(d) for d in DIAS_LABORALES}

    # Fin de la última clase e inicio de la primera, por día laboral.
    # Día sin clases: fin=0 (mejor imposible para "salir temprano") e
    # inicio=DIA_FIN_MIN (mejor imposible para "entrar tarde").
    fin_por_dia = {d: max((fin for _, fin in sesiones_por_dia[d]), default=0)
                   for d in DIAS_LABORALES}
    inicio_por_dia = {d: min((ini for ini, _ in sesiones_por_dia[d]), default=DIA_FIN_MIN)
                      for d in DIAS_LABORALES}

    return {
        "minutos_en_evitar": minutos_en_evitar,
        "dias_con_clase": dias_con_clase,
        "dias_libres": dias_libres,
        "num_dias_con_clase": len(dias_con_clase),
        "usa_sabado": usa_sabado,
        "usa_domingo": usa_domingo,
        "bloques_libres_por_dia": bloques,
        "suma_bloques_libres": sum(bloques.values()),
        "min_bloque_libre": min(bloques.values()),
        "suma_fin_ultima_clase": sum(fin_por_dia.values()),
        "suma_inicio_primera_clase": sum(inicio_por_dia.values()),
    }


def score_manana_compacta(metrics: dict) -> tuple:
    """Salir de clases lo más temprano posible cada día."""
    return (-metrics["minutos_en_evitar"], -metrics["suma_fin_ultima_clase"])


def score_empezar_tarde(metrics: dict) -> tuple:
    """Entrar a clases lo más tarde posible (mañanas libres)."""
    return (-metrics["minutos_en_evitar"], metrics["suma_inicio_primera_clase"])


def score_maximo_dia_libre(metrics: dict) -> tuple:
    """Concentrar las clases en pocos días; días completos sin clase."""
    return (-metrics["minutos_en_evitar"], -metrics["num_dias_con_clase"],
            metrics["suma_bloques_libres"])


def score_bloques_mixtos(metrics: dict) -> tuple:
    """Bloques libres continuos grandes y parejos entre días."""
    return (-metrics["minutos_en_evitar"], metrics["suma_bloques_libres"],
            metrics["min_bloque_libre"])


ESTRATEGIAS = {
    "manana_compacta": {
        "nombre": "Salir temprano",
        "descripcion": "Terminar clases lo antes posible cada día.",
        "score": score_manana_compacta,
    },
    "empezar_tarde": {
        "nombre": "Entrar tarde",
        "descripcion": "Empezar clases lo más tarde posible.",
        "score": score_empezar_tarde,
    },
    "maximo_dia_libre": {
        "nombre": "Máximo día libre",
        "descripcion": "Concentrar clases en pocos días.",
        "score": score_maximo_dia_libre,
    },
    "bloques_mixtos": {
        "nombre": "Bloques libres",
        "descripcion": "Huecos libres grandes y parejos entre días.",
        "score": score_bloques_mixtos,
    },
}


def evaluar(combinaciones: list, bloqueos_evitar: list = ()) -> list:
    """Calcula las métricas una sola vez por combinación -> [(combo, metrics)]."""
    return [(combo, compute_metrics(combo, bloqueos_evitar)) for combo in combinaciones]


def rankear(evaluadas: list, estrategia_id: str, top_n: int = 3) -> list:
    """Ordena [(combo, metrics)] según la estrategia y devuelve las top_n."""
    score = ESTRATEGIAS[estrategia_id]["score"]
    return sorted(evaluadas, key=lambda par: score(par[1]), reverse=True)[:top_n]

File: web/src/lib/engine/strategies.ts
/** Métricas por combinación y puntuación por estrategia.
 * Puerto 1:1 de engine/strategies.py (incluida la semántica de bloqueos:
 * «imposible» se filtra en el solver; «evitar» se minimiza aquí primero). */
import { DIAS_LABORALES, sesionesDe, type Dia, type Sesion } from "./models";
import type { Combo } from "./solver";

/** Franja del día para calcular huecos (lienzo de métricas, no restricción). */
export const DIA_INICIO_MIN = 7 * 60;
export const DIA_FIN_MIN = 19 * 60;

export interface Metrics {
  minutos_en_evitar: number;
  dias_con_clase: Dia[];
  dias_libres: Dia[];
  num_dias_con_clase: number;
  usa_sabado: boolean;
  usa_domingo: boolean;
  bloques_libres_por_dia: Record<string, number>;
  suma_bloques_libres: number;
  min_bloque_libre: number;
  suma_fin_ultima_clase: number;
  suma_inicio_primera_clase: number;
}

function minutosTraslape(sesion: Sesion, bloqueos: Sesion[]): number {
  let total = 0;
  for (const b of bloqueos) {
    const comunes = sesion.dias.filter((d) => b.dias.includes(d)).length;
    if (comunes) {
      const solapa = Math.max(0,
        Math.min(sesion.fin_min, b.fin_min) - Math.max(sesion.inicio_min, b.inicio_min));
      total += solapa * comunes;
    }
  }
  return total;
}

export function computeMetrics(combinacion: Combo, bloqueosEvitar: Sesion[] = []): Metrics {
  const porDia: Record<string, Array<[number, number]>> =
    Object.fromEntries(DIAS_LABORALES.map((d) => [d, []]));
  let usaSabado = false, usaDomingo = false, minutosEnEvitar = 0;

  for (const [, opcion] of combinacion) {
    for (const s of sesionesDe(opcion)) {
      minutosEnEvitar += minutosTraslape(s, bloqueosEvitar);
      for (const d of s.dias) {
        if (porDia[d]) porDia[d].push([s.inicio_min, s.fin_min]);
        else if (d === "SA") usaSabado = true;
        else if (d === "DO") usaDomingo = true;
      }
    }
  }

  const diasConClase = DIAS_LABORALES.filter((d) => porDia[d].length);
  const diasLibres = DIAS_LABORALES.filter((d) => !porDia[d].length);

  const bloqueLibreMasLargo = (dia: Dia): number => {
    const ocupado = [...porDia[dia]].sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    const libres: number[] = [];
    let cursor = DIA_INICIO_MIN;
    for (const [ini, fin] of ocupado) {
      const iniC = Math.max(ini, DIA_INICIO_MIN), finC = Math.min(fin, DIA_FIN_MIN);
      if (iniC > cursor) libres.push(iniC - cursor);
      cursor = Math.max(cursor, finC);
    }
    if (cursor < DIA_FIN_MIN) libres.push(DIA_FIN_MIN - cursor);
    return libres.length ? Math.max(...libres) : DIA_FIN_MIN - DIA_INICIO_MIN;
  };

  const bloques = Object.fromEntries(
    DIAS_LABORALES.map((d) => [d, bloqueLibreMasLargo(d)]));

  const finPorDia = DIAS_LABORALES.map((d) =>
    porDia[d].length ? Math.max(...porDia[d].map(([, f]) => f)) : 0);
  const inicioPorDia = DIAS_LABORALES.map((d) =>
    porDia[d].length ? Math.min(...porDia[d].map(([i]) => i)) : DIA_FIN_MIN);

  return {
    minutos_en_evitar: minutosEnEvitar,
    dias_con_clase: diasConClase,
    dias_libres: diasLibres,
    num_dias_con_clase: diasConClase.length,
    usa_sabado: usaSabado,
    usa_domingo: usaDomingo,
    bloques_libres_por_dia: bloques,
    suma_bloques_libres: Object.values(bloques).reduce((a, b) => a + b, 0),
    min_bloque_libre: Math.min(...Object.values(bloques)),
    suma_fin_ultima_clase: finPorDia.reduce((a, b) => a + b, 0),
    suma_inicio_primera_clase: inicioPorDia.reduce((a, b) => a + b, 0),
  };
}

/** Todas las llaves son "mayor es mejor"; el primer criterio SIEMPRE es
 * minimizar los minutos en zonas «mejor no». */
type Score = (m: Metrics) => number[];

export const ESTRATEGIAS: Record<string, { nombre: string; descripcion: string; score: Score }> = {
  manana_compacta: {
    nombre: "Salir temprano",
    descripcion: "Terminar clases lo antes posible cada día.",
    score: (m) => [-m.minutos_en_evitar, -m.suma_fin_ultima_clase],
  },
  empezar_tarde: {
    nombre: "Entrar tarde",
    descripcion: "Empezar clases lo más tarde posible.",
    score: (m) => [-m.minutos_en_evitar, m.suma_inicio_primera_clase],
  },
  maximo_dia_libre: {
    nombre: "Máximo día libre",
    descripcion: "Concentrar clases en pocos días.",
    score: (m) => [-m.minutos_en_evitar, -m.num_dias_con_clase, m.suma_bloques_libres],
  },
  bloques_mixtos: {
    nombre: "Bloques libres",
    descripcion: "Huecos libres grandes y parejos entre días.",
    score: (m) => [-m.minutos_en_evitar, m.suma_bloques_libres, m.min_bloque_libre],
  },
};

export type Evaluada = [Combo, Metrics];

export function evaluar(combinaciones: Combo[], bloqueosEvitar: Sesion[] = []): Evaluada[] {
  return combinaciones.map((c) => [c, computeMetrics(c, bloqueosEvitar)]);
}

/** Orden descendente estable por la tupla de score (== sorted(reverse=True)). */
export function rankear(evaluadas: Evaluada[], estrategiaId: string, topN = 3): Evaluada[] {
  const score = ESTRATEGIAS[estrategiaId].score;
  return [...evaluadas]
    .map((e, i) => ({ e, i, s: score(e[1]) }))
    .sort((a, b) => {
      for (let k = 0; k < Math.max(a.s.length, b.s.length); k++) {
        const d = (b.s[k] ?? 0) - (a.s[k] ?? 0);
        if (d) return d;
      }
      return a.i - b.i;   // estabilidad explícita
    })
    .slice(0, topN)
    .map(({ e }) => e);
}
